[Completion] Remove uses of unbounded generic types

hamishknight opened the pull request:
Ensure we always produce bounded generic type annotations and contextual types. While here, also make sure we substitute the base type for nested type completions.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented on test/IDE/complete_member_type.swift:
```diff
@@ -115,10 +115,10 @@ do {
-  // DICTIONARY-NEXT: Decl[Struct]/CurrNominal/IsSystem: Values[#Dictionary.Values#]; name=Values
-  // DICTIONARY-NEXT: Decl[Struct]/CurrNominal/IsSystem: Index[#Dictionary.Index#]; name=Index
-  // DICTIONARY-NEXT: Decl[Struct]/CurrNominal/IsSystem: Iterator[#Dictionary.Iterator#]; name=Iterator
+  // DICTIONARY-NEXT: Decl[Struct]/CurrNominal/IsSystem: Keys[#Dictionary<Key, Value>.Keys#]; name=Keys
+  // DICTIONARY-NEXT: Decl[Struct]/CurrNominal/IsSystem: Values[#Dictionary<Key, Value>.Values#]; name=Values
+  // DICTIONARY-NEXT: Decl[Struct]/CurrNominal/IsSystem: Index[#Dictionary<Key, Value>.Index#]; name=Index
+  // DICTIONARY-NEXT: Decl[Struct]/CurrNominal/IsSystem: Iterator[#Dictionary<Key, Value>.Iterator#]; name=Iterator
```
We really ought to be substituting the base type for these

slavapestov commented on lib/IDE/CodeCompletionResultType.cpp:
```diff
@@ -395,8 +395,7 @@ static TypeRelation calculateTypeRelation(Type Ty, Type ExpectedTy,
   // requirements – ignore them
-  if (!Ty->hasTypeParameter() && !ExpectedTy->hasTypeParameter() &&
-      !Ty->hasUnboundGenericType() && !ExpectedTy->hasUnboundGenericType()) {
+  if (!Ty->hasTypeParameter() && !ExpectedTy->hasTypeParameter()) {
```
Can we also get rid of type parameters at some point too, in favor of the caller mapping the type into the correct environment?

hamishknight replied:
Yup I'm leaving that for a future PR

slavapestov reviewed (approved):
Thanks!

hamishknight committed:
[Completion] Remove uses of unbounded generic types

Ensure we always produce bounded generic type annotations and
contextual types.

hamishknight committed:
[Completion] NFC: Factor out `getMemberBaseType()`

hamishknight committed:
[Completion] Don't suggest nested type in where clause

We allow the unqualifed use of the enclosing nominal type in a where
clause, but only when it isn't a nested type, e.g:

```
struct S<T> {
  typealias T = T
  struct R<U> {
    typealias U = U
  }
}
extension S where S.T == Int {}   // allowed
extension S.R where R.U == Int {} // not allowed
```

Tweak the completion logic such that we don't suggest the type for
the nested case, instead it must be qualified.

hamishknight committed:
[Completion] Substitute base type for nested types

Ensure we call `getTypeOfMember` for nested nominals and typealiases.

hamishknight commented:
@swift-ci please SourceKit stress test

ahoppen reviewed (approved):
LGTM. Some nice improvements in the tests as well 😄

hamishknight commented:
@swift-ci please SourceKit stress test
